### PR TITLE
Use VerifyAttestation to Filter Attestation

### PR DESCRIPTION
**Pull request by terenc3t**
As specified in the [validator spec](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/validator/0_beacon-chain-validator.md#attestations) when a proposer includes an attestation it should run `process_attestation` to make sure the attestation passes verification before inclusion.

In our case we should run `VerifyAttestation` before including the attestation to the beacon block

### CODE REVIEW

File: beacon-chain/rpc/proposer_server.go [hunk 1]
```diff
@@ -119,16 +120,20 @@ func (ps *ProposerServer) PendingAttestations(ctx context.Context, req *pb.Pendi
...
 		if helpers.SlotToEpoch(att.Data.Slot+1) >= helpers.SlotToEpoch(currentSlot) {
 			expectedJustifedEpoch = beaconState.JustifiedEpoch
 		} else {
 			expectedJustifedEpoch = beaconState.PreviousJustifiedEpoch
 		}
 
-		if att.Data.Slot > boundary &&
-			att.Data.JustifiedEpoch == expectedJustifedEpoch &&
-			ps.chainService.IsCanonical(att.Data.Slot, att.Data.BeaconBlockRootHash32) {
+		if att.Data.Slot > boundary && att.Data.JustifiedEpoch == expectedJustifedEpoch {
```
Comment (prestonvanloon) [inline beacon-chain/rpc/proposer_server.go]: Do we need this if verify attestation handles all of the validity checks?

File: beacon-chain/rpc/proposer_server_test.go [hunk 2]
```diff
@@ -198,17 +200,17 @@ func TestPendingAttestations_FiltersExpiredAttestations(t *testing.T) {
...
-			{Data: &pbp2p.AttestationData{Slot: currentSlot - 100, JustifiedEpoch: expectedEpoch}},
-			{Data: &pbp2p.AttestationData{Slot: currentSlot - params.BeaconConfig().SlotsPerEpoch, JustifiedEpoch: expectedEpoch}},
-			// Non-expired attestation with incorrect justified epoch
-			{Data: &pbp2p.AttestationData{Slot: currentSlot - 5, JustifiedEpoch: expectedEpoch - 1}},
+			//{Data: &pbp2p.AttestationData{Slot: 0, JustifiedEpoch: expectedEpoch, CrosslinkDataRootHash32: params.BeaconConfig().ZeroHash[:],}},
+			//{Data: &pbp2p.AttestationData{Slot: currentSlot - 10000, JustifiedEpoch: expectedEpoch, CrosslinkDataRootHash32: params.BeaconConfig().ZeroHash[:],}},
+			//{Data: &pbp2p.AttestationData{Slot: currentSlot - 5000, JustifiedEpoch: expectedEpoch, CrosslinkDataRootHash32: params.BeaconConfig().ZeroHash[:],}},
+			//{Data: &pbp2p.AttestationData{Slot: currentSlot - 100, JustifiedEpoch: expectedEpoch, CrosslinkDataRootHash32: params.BeaconConfig().ZeroHash[:],}},
+			//{Data: &pbp2p.AttestationData{Slot: currentSlot - params.BeaconConfig().SlotsPerEpoch, JustifiedEpoch: expectedEpoch, CrosslinkDataRootHash32: params.BeaconConfig().ZeroHash[:],}},
```
Comment (prestonvanloon) [inline beacon-chain/rpc/proposer_server_test.go]: ?

Comment (terenc3t) [reply beacon-chain/rpc/proposer_server_test.go]: sorry, fixing that now

File: beacon-chain/rpc/proposer_server.go (hunk 1 above)
Comment (terenc3t) [reply beacon-chain/rpc/proposer_server.go]: We don't, I'll remove it. `VerifyAttestation` handles those 2 checks already

Comment (rauljordan) [reply beacon-chain/rpc/proposer_server.go]: Good point

File: beacon-chain/rpc/proposer_server.go [hunk 3]
```diff
@@ -110,29 +111,17 @@ func (ps *ProposerServer) PendingAttestations(ctx context.Context, req *pb.Pendi
...
 	for _, att := range atts {
-
-		var expectedJustifedEpoch uint64
-		if helpers.SlotToEpoch(att.Data.Slot+1) >= helpers.SlotToEpoch(currentSlot) {
-			expectedJustifedEpoch = beaconState.JustifiedEpoch
-		} else {
-			expectedJustifedEpoch = beaconState.PreviousJustifiedEpoch
+		if err := blocks.VerifyAttestation(beaconState, att, false); err != nil {
+			log.WithField("slot", att.Data.Slot-
+				params.BeaconConfig().GenesisSlot).Warn("Skipping, pending attestation failed verification")
```
Comment (prestonvanloon) [inline beacon-chain/rpc/proposer_server.go]: Do you want to log WithError(err)?

Comment (terenc3t) [reply beacon-chain/rpc/proposer_server.go]: yeah... error feels more appropriate here

**Pull request merged**